flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/ZooKeeperSubmittedJobGraphs.java

```diff
@@ -0,0 +1,356 @@
+
+			SubmittedJobGraph jobGraph = jobStateHandle
+					.getState(ClassLoader.getSystemClassLoader());
+
+			addedJobGraphs.add(jobGraph.getJobId());
+
+			LOG.info("Recovered {}.", jobGraph);
+			return jobGraph;
+		}
+	}
+
+	@Override
+	public void putJobGraph(SubmittedJobGraph jobGraph) throws Exception {
+		checkNotNull(jobGraph, "Job graph");
+		String path = getPathForJob(jobGraph.getJobId());
+
+		synchronized (cacheLock) {
+			verifyIsRunning();
+
+			int currentVersion = jobGraphsInZooKeeper.exists(path);
+
+			if (currentVersion == -1) {
+				jobGraphsInZooKeeper.add(path, jobGraph);
+				LOG.info("Added {} to ZooKeeper.", jobGraph);
+			}
+			else if (addedJobGraphs.contains(jobGraph.getJobId())) {
+				jobGraphsInZooKeeper.replace(path, currentVersion, jobGraph);
+				LOG.info("Updated {} in ZooKeeper.", jobGraph);
+			}
+			else {
+				throw new IllegalStateException("Oh, no. Trying to update a graph you didn't " +
+						"#getAllSubmittedJobGraphs() or #putJobGraph() yourself before.");
+			}
+
+			addedJobGraphs.add(jobGraph.getJobId());
+		}
+	}
+
+	@Override
+	public void removeJobGraph(JobID jobId) throws Exception {
+		checkNotNull(jobId, "Job ID");
+		String path = getPathForJob(jobId);
+
+		synchronized (cacheLock) {
+			if (addedJobGraphs.contains(jobId)) {
+				jobGraphsInZooKeeper.discard(path);
+
+				addedJobGraphs.remove(jobId);
+				LOG.info("Removed job graph {} from ZooKeeper.", jobId);
+			}
+		}
+	}
+
+	/**
+	 * Monitors ZooKeeper for changes.
+	 *
+	 * <p>Detects modifications from other job managers in corner situations. The event
+	 * notifications fire for changes from this job manager as well.
+	 */
+	private final class SubmittedJobGraphsPathCacheListener implements PathChildrenCacheListener {
+
+		@Override
+		public void childEvent(CuratorFramework client, PathChildrenCacheEvent event)
+				throws Exception {
+
+			if (LOG.isDebugEnabled()) {
+				if (event.getData() != null) {
+					LOG.debug("Received {} event (path: {})", event.getType(), event.getData().getPath());
+				}
+				else {
+					LOG.debug("Received {} event", event.getType());
+				}
+			}
+
+			synchronized (cacheLock) {
+				switch (event.getType()) {
+					case CHILD_ADDED:
+						try {
+							JobID jobId = fromEvent(event);
+							if (jobGraphListener != null && !addedJobGraphs.contains(jobId)) {
+								try {
+									// Whoa! This has been added by someone else. Or we were fast
+									// to remove it (false positive).
+									jobGraphListener.onAddedJobGraph(jobId);
+								}
+								catch (Throwable t) {
+									LOG.error("Error in callback", t);
+								}
+							}
+						}
+						catch (Exception e) {
+							LOG.error("Error in SubmittedJobGraphsPathCacheListener", e);
+						}
+
+						break;
+
+					case CHILD_UPDATED:
+						// TODO What to do if someone overwrites one of our job graphs? And how
+						// to tell whose it is? Is the update-time check enough to prevent these
+						// problematic situations.
```

Comment:
Resolved. I don't think that this a case we have to monitor atm. Removed the TODO.